### PR TITLE
`LOG_CONNECTOR_MESSAGES` enables logging of all messages seen by the worker in a sync

### PR DESCRIPTION
**Pull request by evantahler**
As an Airbyte developer, I often want to inspect the literal JSON messages passing though the system between source and destination.  This PR adds a new environment variable feature flag, `LOG_CONNECTOR_MESSAGES` which can be used to toggle this on.  

For example: `LOG_CONNECTOR_MESSAGES=true VERSION=dev docker-compose up`

When enabled, messages will be logged like:

```
airbyte-worker         | 2022-08-23 22:07:01 INFO i.a.w.i.AirbyteMessageTracker(logMessageAsJSON):402 - source message | {"type":"RECORD","record":{"stream":"Purchases","data":{"id":480,"product_id":56,"user_id":400,"added_to_cart_at":"2017-04-11T13:32:12","purchased_at":"2017-09-16T13:32:12","returned_at":null},"emitted_at":1661292420000}}
airbyte-worker         | 2022-08-23 22:07:01 INFO i.a.w.i.AirbyteMessageTracker(logMessageAsJSON):402 - source message | {"type":"STATE","state":{"data":{"Products":{"product_count":100},"Users":{"cursor":400,"seed":-1}}}}
airbyte-worker         | 2022-08-23 22:07:01 INFO i.a.w.i.AirbyteMessageTracker(logMessageAsJSON):402 - destination message | {"type":"STATE","state":{"data":{"Products":{"product_count":100},"Users":{"cursor":500,"seed":-1},"Purchases":{"purchases_count":600}}}}
```

This PR also adds logging for the source's starting state as passed in via the JSON file so the full picture of the source's lifecycle can be determined:

```
airbyte-worker         | 2022-08-23 23:37:14 INFO i.a.w.i.DefaultAirbyteSource(logInitialStateAsJSON):167 -  source starting state | {"state":[{"type":"STREAM","stream":{"stream_state":{"cursor":"600.0","stream_name":"purchases","cursor_field":["id"],"stream_namespace":"public"},"stream_descriptor":{"name":"purchases","namespace":"public"}}},{"type":"STREAM","stream":{"stream_state":{"cursor":"500.0","stream_name":"users","cursor_field":["id"],"stream_namespace":"public"},"stream_descriptor":{"name":"users","namespace":"public"}}},{"type":"STREAM","stream":{"stream_state":{"cursor":"100.0","stream_name":"products","cursor_field":["id"],"stream_namespace":"public"},"stream_descriptor":{"name":"products","namespace":"public"}}}]}
```

### CODE REVIEW

**Comment by evantahler** [comment]
@brianjlai I added logging for the initial state passed to the connector as well (See description).

**Review by evantahler — COMMENTED** [review]
Questions for PR reviewers:
* Where else does the new ENV variable need to go?
* Where should I document this new ENV variable?

File: airbyte-workers/src/main/java/io/airbyte/workers/internal/AirbyteMessageTracker.java [hunk 1]
```diff
@@ -386,4 +393,17 @@ protected Long calculateMean(final Long currentMean, final Long totalCount, fina
     return (long) result;
   }
 
+  private void logMessageAsJSON(final String caller, AirbyteMessage message) {
+    if (!logConnectorMessages) {
+      return;
+    }
+
+    try {
+      final String json = new ObjectMapper().writeValueAsString(message);
```
Comment (gosusnp) [inline airbyte-workers/src/main/java/io/airbyte/workers/internal/AirbyteMessageTracker.java:402]: I think we have [a helper for that](https://github.com/airbytehq/airbyte/blob/e9afa9bef3fcd5bf4237bea1295973ccd0c1f1e5/airbyte-commons/src/main/java/io/airbyte/commons/json/Jsons.java#L44)

File: airbyte-workers/src/main/java/io/airbyte/workers/internal/DefaultAirbyteSource.java [hunk 2]
```diff
@@ -150,4 +156,17 @@ public void cancel() throws Exception {
     }
   }
 
+  private void logInitialStateAsJSON(final WorkerSourceConfig sourceConfig) {
+    if (!logConnectorMessages) {
+      return;
+    }
+
+    try {
+      final String json = new ObjectMapper().writeValueAsString(sourceConfig.getState().getState());
```
Comment (gosusnp) [inline airbyte-workers/src/main/java/io/airbyte/workers/internal/DefaultAirbyteSource.java:165]: Could use the helper here as well.

**Review by gosusnp — APPROVED** [review]
LGTM, I think we could use the common helper here for consistency.
Any docs need to be updated so that people have a better chance of finding this feature?

File: airbyte-workers/src/main/java/io/airbyte/workers/internal/DefaultAirbyteSource.java [hunk 3]
```diff
@@ -150,4 +156,17 @@ public void cancel() throws Exception {
...
+      LOGGER.info("source starting state | " + json);
+    } catch (JsonProcessingException e) {
+      LOGGER.warn("Error serializing " + sourceConfig.getState() + " to JSON: " + e);
```
Comment (brianjlai) [inline airbyte-workers/src/main/java/io/airbyte/workers/internal/DefaultAirbyteSource.java:168]: should this also be `sourceConfig.getState().getState()` same as above?

File: airbyte-workers/src/main/java/io/airbyte/workers/internal/DefaultAirbyteSource.java (hunk 2 above)
Comment (brianjlai) [reply airbyte-workers/src/main/java/io/airbyte/workers/internal/DefaultAirbyteSource.java:165]: is there somewhere to confirm that we do this getState() twice before we pass it to the connector? This seems to lineup with what I see in the CDK where stream names are at the top level of the state JSON blob. But this is a pretty critical part of the platform <> connector interface

**Review by brianjlai — APPROVED** [review]
lgtm this will be great help for development and debugging!

**Comment by gosusnp** [comment]
Saw that afterwards

> * Where else does the new ENV variable need to go?
As long as we do not want to propagate this to the connectors, we should be good. Otherwise there is an extra step for that.

> * Where should I document this new ENV variable?
What about [developing-locally.md](https://github.com/airbytehq/airbyte/blob/e9afa9bef3fcd5bf4237bea1295973ccd0c1f1e5/docs/contributing-to-airbyte/developing-locally.md)?

File: airbyte-workers/src/main/java/io/airbyte/workers/internal/DefaultAirbyteSource.java (hunk 3 above)
Comment (evantahler) [reply airbyte-workers/src/main/java/io/airbyte/workers/internal/DefaultAirbyteSource.java]: removed in https://github.com/airbytehq/airbyte/pull/15895/commits/3b0006f7347a7f6115a08d0784fc64345ae2bbc7

File: airbyte-workers/src/main/java/io/airbyte/workers/internal/AirbyteMessageTracker.java (hunk 1 above)
Comment (evantahler) [reply airbyte-workers/src/main/java/io/airbyte/workers/internal/AirbyteMessageTracker.java]: Nice! added via https://github.com/airbytehq/airbyte/pull/15895/commits/3b0006f7347a7f6115a08d0784fc64345ae2bbc7

**Comment by gosusnp** [comment]
About @brianjlai 's question, I wonder if it's worth capturing the result of [this call](https://github.com/airbytehq/airbyte/blob/3894134d1115b7c9ba181684f1aba66bec500a63/airbyte-workers/src/main/java/io/airbyte/workers/internal/DefaultAirbyteSource.java#L78) rather than re-accessing the sourceConfig, that would probably be more consistent and less chance to be broken.

**Pull request merged**